src/devices/SM28BYJ48/SM28BYJ48.cs

```diff
@@ -0,0 +1,202 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Device.Gpio;
+using System.Threading;
+
+namespace Iot.Device.SM28BYJ48
+{
+    /// <summary>
+    /// Can move the motor clockwise and counterclockwise.
+    /// </summary>
+    public enum RotationDirection
+    {
+        Stopped,
+        Clockwise,
+        CounterClockwise
+    }
+
+    /// <summary>
+    /// The 28BYJ-48 motor has 512 full engine rotations to rotate the drive shaft once.
+    /// In half-step mode these are 8 x 512 = 4096 steps for a full rotation, and in full-step mode these are
+    /// 4 x 512 = 2048 steps for a full rotation.
+    /// </summary>
+    public enum StepperType
+    {
+        HalfStep,
+        FullStepSinglePhase,
+        FullStepDualPhase
+    }
+
+    /// <summary>
+    /// This class is for controlling stepper motors that are controlled by a 4 pin controller board.
+    /// </summary>
+    /// <remarks>It is tested and developed using the 28BYJ-48 stepper motor and the ULN2003 driver board.</remarks>
+    public class SM28BYJ48 : IDisposable
+    {
+        private int _pin1;
+        private int _pin2;
+        private int _pin3;
+        private int _pin4;
+        private int _delayMs = 1;
+        private int _steps = 0;
+        private int _engineStep = 0;
+        private int _currentStep = 0;
+        private StepperType _type = StepperType.HalfStep;
+        private RotationDirection _rotationDirection = RotationDirection.Stopped;
+        private GpioController _controller;
+        private Action _movementCompleted;
+
+        /// <summary>
+        /// Initialize a StepperMotor class
+        /// </summary>
+        /// <param name="pin1">The GPIO pin number which corresponds pin A on ULN2003 driver board.</param>
+        /// <param name="pin2">The GPIO pin number which corresponds pin B on ULN2003 driver board.</param>
+        /// <param name="pin3">The GPIO pin number which corresponds pin C on ULN2003 driver board.</param>
+        /// <param name="pin4">The GPIO pin number which corresponds pin D on ULN2003 driver board.</param>
+        public SM28BYJ48(int pin1, int pin2, int pin3, int pin4)
```

Comment:
do all stepper have 4 windings or can they have i.e. 8 for more precision? (This is fine as is for now regardless of the answer since we can easily change it later)

Reply:
Unipolar stepper motors (like 28BYJ-48) have five or six contacts for connection and four electromagnetic coils in the housing (to be more precise, two coils divided into four). I found more info [here](https://learn.adafruit.com/all-about-stepper-motors/types-of-steppers)